cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -214,98 +193,223 @@ public PipelinePlan plan(PipelineSpec spec) {
+    // when two contol nodes are chained together for example Action-->Condition. We need to add connection
+    // between them. Also we need to process any control nodes which are NOT yet in the processedControlNodes set.
     for (Map.Entry<String, Dag> dagEntry : subdags.entrySet()) {
-      // If dag only contains conditions then ignore it
-      if (conditionBranches.keySet().containsAll(dagEntry.getValue().getNodes())) {
+      Dag dag = dagEntry.getValue();
+      if (controlNodes.containsAll(dag.getSources())) {
+        String source = dag.getSources().iterator().next();
+        String sink = dagEntry.getKey();
+        if (controlNodes.containsAll(dagEntry.getValue().getNodes())) {
+          // Current dag only contain control nodes.
+          sink = dagEntry.getValue().getSinks().iterator().next();
+        }
+        if (!processedControlNodes.contains(source)) {
+          // This control node is not processed yet, because it is first in the pipeline.
+          ConditionBranches branches = conditionBranches.get(source);
+          if (branches != null) {
+            // this is condition node
+            boolean condition = false;
+            if (branches.getTrueOutput() != null && dag.getNodes().contains(branches.getTrueOutput())) {
+              condition = true;
+            }
+            phaseConnections.add(new Connection(source, sink, condition));
+          } else {
+            phaseConnections.add(new Connection(source, sink));
+          }
+          processedControlNodes.add(source);
+        }
+      }
+
+      if (controlNodes.containsAll(dag.getSinks())) {
+        String source = dagEntry.getKey();
+        if (controlNodes.containsAll(dagEntry.getValue().getNodes())) {
+          source = dagEntry.getValue().getSources().iterator().next();
+        }
+        String sink = dag.getSinks().iterator().next();
+        if (!processedControlNodes.contains(sink)) {
+          // This control node is not processed yet, because it is last in the pipeline.
+          // Conditions cannot be last however action nodes can be
+          phaseConnections.add(new Connection(source, sink));
+          processedControlNodes.add(sink);
+        }
+      }
+
+      // If dag only contains control nodes then ignore it, since all of its connections are processed already
+      if (controlNodes.containsAll(dagEntry.getValue().getNodes())) {
         continue;
       }
-      Dag dag = getUpdatedDag(dagEntry.getValue(), conditionConnectors);
-      phases.put(dagEntry.getKey(), dagToPipeline(dag, connectorNodes, specs, conditionConnectors));
-    }
 
-    populateActionPhases(specs, actionNodes, phases, phaseConnections, outgoingActionConnections,
-                         incomingActionConnections, subdags);
+      Dag updatedDag = getUpdatedDag(dagEntry.getValue(), controlConnectors);
+      // Remove any action nodes from this dag
```

Comment:
since we split based on the control stages this would result in multiple dags <action, contdition>, <contdition, action>, <action, source...>. Now the dags which only contain control nodes (<action, condition>, <conditon, action>) we filter them out since corresponding phases and phase connections are added at this point. Now remaining dag is <action, source...>. since action is already added as a separate phase, we need to remove any control stage from the dag before creating phase out of it.